controllers/resources/lease.go

```diff
@@ -0,0 +1,233 @@
+	//ManageLease extends or releases a lease based on the CR status, type of remediation and how long the lease is already leased
+	ManageLease(ctx context.Context, remediationCR *unstructured.Unstructured, nhc *remediationv1alpha1.NodeHealthCheck) (time.Duration, error)
+	// InvalidateLease extends or releases a lease based on the CR status, type of remediation and how long the lease is already leased
+	InvalidateLease(ctx context.Context, remediationCR *unstructured.Unstructured) error
+}
+
+type nhcLeaseManager struct {
+	client             client.Client
+	commonLeaseManager lease.Manager
+	log                logr.Logger
+}
+
+func NewLeaseManager(client client.Client, log logr.Logger) (LeaseManager, error) {
+	newManager, err := lease.NewManager(client, holderIdentity)
+	if err != nil {
+		log.Error(err, "couldn't initialize lease manager")
+		return nil, err
+	}
+	return &nhcLeaseManager{
+		client:             client,
+		commonLeaseManager: newManager,
+		log:                log.WithName("nhc lease manager"),
+	}, nil
+}
+
+func (m *nhcLeaseManager) ObtainNodeLease(remediationCR *unstructured.Unstructured, nhc *remediationv1alpha1.NodeHealthCheck) (bool, *time.Duration, error) {
+	nodeName := remediationCR.GetName()
+	leaseDuration := m.getLeaseDurationForRemediation(remediationCR, nhc)
+	leaseDurationWithBuffer := leaseDuration + LeaseBuffer
+
+	node := &v1.Node{}
+	if err := m.client.Get(context.Background(), types.NamespacedName{Name: nodeName}, node); err != nil {
+		m.log.Error(err, "couldn't obtain node lease node error getting node", "node name", nodeName)
+		return false, nil, err
+	}
+
+	if err := m.commonLeaseManager.RequestLease(context.Background(), node, leaseDurationWithBuffer); err != nil {
+		if _, ok := err.(lease.AlreadyHeldError); ok {
+			m.log.Info("can't acquire node lease, it is already owned by another owner", "already held error", err)
+			return false, &RequeueIfLeaseTaken, err
+		}
+
+		m.log.Error(err, "couldn't obtain lease for node", "node name", nodeName)
+		return false, nil, err
+	}
+
+	//all good lease created with wanted duration
+	return true, &leaseDuration, nil
+
+}
+
+func (m *nhcLeaseManager) ManageLease(ctx context.Context, remediationCR *unstructured.Unstructured, nhc *remediationv1alpha1.NodeHealthCheck) (time.Duration, error) {
+	node := &v1.Node{}
+	if err := m.client.Get(ctx, client.ObjectKey{Name: remediationCR.GetName()}, node); err != nil {
+		m.log.Error(err, "managing lease - couldn't fetch node", "node name", remediationCR.GetName())
+		return 0, err
+	}
+	l, err := m.commonLeaseManager.GetLease(ctx, node)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			return 0, nil
+		}
+		m.log.Error(err, "managing lease - couldn't fetch lease", "node name", remediationCR.GetName())
+		return 0, err
+	}
+	exist := remediationCR.GetDeletionTimestamp() == nil
+	if !exist && m.isLeaseOwner(l) {
```

Comment:
> The !exist doesn't check the lease, it checks whether the remediation is about to be deleted

Oh sorry, I misread the code 🙈 

> ManageLease is called only after a remediation CR is created, and the condition to creating a Remediation CR is owning the lease.

Anyone can change leases. You shouldn't update/delete not owned leases, but no one is preventing it. So IMHO better safe than sorry, let's not be that bad guy ourself, and always check ownership IMHO.

Reply:
Well consider the following scenario:
1. Unhealthy Node detected
2. Obtain lease creates a lease
3. Remediation is created
4. Node stays unhealthy for a long time (remediation is unsuccessful)
5. Lease is deleted because it's overdue.

Assuming we allow `ManageLease` to create leases: at that point (in case a remediation occur) we will create a new lease for that remediation, which would be wrong IMO.

It is true that technically active leases can be deleted by anyone, but that would "break the lease contract" so I think it's a reasonable assumption to assume this "contract" is upheld - I prefer this approach over the alternative I've described.

Reply:
>  at that point (in case a remediation occur) we will create a new lease for that remediation, which would be wrong IMO.

yes, that would be wrong, but I think that won't happen because we won't try to create a new remediation CR / lease for a timed out remediation

> Coming to think of it NMO original implementation of invalidating the lease didn't delete the lease, maybe it was so to address this type of use-case

might be... but leaving leases forever also is wrong, not? 🤔

Reply:
>yes, that would be wrong, but I think that won't happen because we won't try to create a new remediation CR / lease for a timed out remediation

We don't have to.
IIUC Any triggering to `Reconcile` (for example creating a remediation for another node) will cause that scenario since `ManageLease` is triggered for every unhealthy node.

>might be... but leaving leases forever also is wrong, not? thinking

That's a really good question.
On the hand it's removing the leases is definitely better way for cleanup & resource management, but on the other hand I don't think it's a huge issue leaving the lease since we still fulfill the "lease contract"  by "closing" it even though it's not deleted.

Thinking from the top of my mind: maybe a better solution will be to do the deletion of the lease only when the CR is removed :thinking:  (Not in this PR though :smile: )

Reply:
> IIUC Any triggering to Reconcile (for example creating a remediation for another node) will cause that scenario since ManageLease is triggered for every unhealthy node.

Ah yes, correct for the single remediation. Oh, and when setting timeout annotation / status update fails...

So the correct flow would be:

- in ManageLease:
  - for existing CR without timeout annotation:
    - when lease overdue: do not invalidate yet, just return OverDue error
    - when lease not found: create new
  - for existing CRs with timeout annotation: invalidate lease and return OverDue error (or a new error)
- in reconciler, when OverDue error occurs:
  - if timeout annotation does not exist: add it and requeue
  - if timeout annotation does exist:  done

Does that make sense? 🤔 

But we can implement this in a follow up, and skip creating new leases in ManageLease for now.

Reply:
>Ah yes, correct for the single remediation.

Why only single ? IIUC it also apply for escalating remediation - IIRC we do not delete the CR in escalating remediation so it will still be detected as unhealthy.

>But we can implement this in a follow up, and skip creating new leases in ManageLease for now.

IIUC this discussion is about better handling the use-case where someone has deleted a valid lease, I agree it's niche enough to discuss later on.

Reply:
hm, actually I think we need to apply parts of the flow, even if we don't want to create new leases:

when a lease is overdue and invalidated, but the remediation CR (timeout annotation) or NHC CR (status) update fails, these updates will not be tried again, because in the next reconcile ManageLease doesn't return the OverDue error again... 🤔

Reply:
> Why only single ?

because for escalating remediation it won't find an appropriate template which for which we doesn't have a timed out CR already, so it stops reconcile before trying to create a remediation CR

Reply:
> because in the next reconcile ManageLease doesn't return the OverDue error again... thinking

IIUC the TimeOutAnnotation will still be placed since it's checked independently of the overdue error [here](https://github.com/medik8s/node-healthcheck-operator/pull/212/files#diff-42b14dc2a70877933db4647434fcbe9b65a8859c7ec1911b0c9da7361c8e98b1R571).

>because for escalating remediation it won't find an appropriate template which for which we don't have a timed out CR already, so it stops reconcile before trying to create a remediation CR


Going back to this scenario: (let's say it happens for node A)
>1. Unhealthy Node detected
>2. Obtain lease creates a lease
>3. Remediation is created 
>4. Node stays unhealthy for a long time (remediation is unsuccessful)
>5. Lease is deleted because it's overdue.

- I think we can agree 1-5 can happen both in escalation / non escalation.
- Now let's say Node B becomes unhealthy and reconcile is triggered.
- we iterate over all existing remediation (in all nodes)
- Node A still has an existing remediation (might be even more than one)
- assuming manage lease also creates leases
- we will try to create a lease for node A, even though it already had a lease that was deleted 
- In this case a lease is created wrongfully because of a remediation that was created by escalation remediation 

Am I missing something here ?

Reply:
> IIUC the TimeOutAnnotation will still be placed since it's checked independently of the overdue error

hm, fair point, because overdue and timeout are directly related

> we will try to create a lease for node A, even though it already had a lease that was deleted

no, because `ManageLease()` won't be called for node A. When the lease was overdue, all possible remediations timed out, and `GetCurrentTemplateWithTimeout()` returns a `NoTemplateLeftError`, and stops remediation for that node.

Reply:
>all possible remediations timed out, and GetCurrentTemplateWithTimeout() returns a NoTemplateLeftError

Cool, thanks :+1:  - I was missing that part.